### PR TITLE
[HttpFoundation] add websocket support

### PR DESCRIPTION
**Pull request by alekitto**
As stated in the subject, this PR adds websocket support in the HttpFoundation and HttpServer components.

To support websocket communcations the user must extend the `Websocket` class and implement its own `onMessage` and optionally `onClose` methods.

* `onMessage` will be invoked when a message has been completely received from the client and a `Message` object is passed as the first parameter.
* `onClose` will be invoked **just before** a close frame is sent to the client.

The `send` method must be used to push a message to the client, passing an instance of `Message` to be sent.

TODO:

- [x] Name changes `Frame` -> `FrameBuffer`, `Packet` -> `Frame` to be coherent with the spec

### CODE REVIEW

File: src/Component/HttpFoundation/src/Websocket/Frame.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
+/**
+ * @memberOf Jymfony.Component.HttpFoundation.Websocket
+ */
+class Frame {
+    __construct() {
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/Frame.js:8]: jsdoc

File: src/Component/HttpFoundation/src/Websocket/FrameBuffer.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+const Message = Jymfony.Component.HttpFoundation.Websocket.Message;
+
+/**
+ * Utility class to decode a websocket frame.
+ *
+ * @memberOf Jymfony.Component.HttpFoundation.Websocket
+ * @internal
+ */
+class FrameBuffer {
+    __construct() {
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/FrameBuffer.js:16]: jsdoc

File: src/Component/Debug/src/Exception/UnhandledRejectionException.js [hunk 3]
```diff
@@ -11,7 +11,7 @@ class UnhandledRejectionException extends RuntimeException {
      * @param {Error} previous
      */
     __construct(promise, previous = undefined) {
-        super.__construct('Unhandled rejection', null, previous);
+        super.__construct('Unhandled rejection: ' + (previous ? previous.toString() : ''), null, previous);
```
Comment (massimilianobraglia) [inline src/Component/Debug/src/Exception/UnhandledRejectionException.js]: If previous exception is undefined, should be a different message shown to the user?
in this way you the message `Unhandled rejection: ` without any explanation will be the exception message.

File: src/Component/HttpFoundation/src/Websocket/FrameBuffer.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Adds a frame to the buffer.
+     *
+     * @param {Jymfony.Component.HttpFoundation.Websocket.Frame} frame
+     *
+     * @private
+     */
+    _addFrame(frame) {
+        if (0 < this._frames.length && 0 !== frame.opcode) {
+            throw new Exception('Bad frame received. Closing connection.');
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/FrameBuffer.js]: A specific exception should be used. Something like `BadFrameReceivedException`.

File: src/Component/HttpFoundation/src/Websocket/Message.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
+/**
+ * @memberOf Jymfony.Component.HttpFoundation.Websocket
+ */
+class Message {
+    __construct(data, type = undefined) {
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/Message.js:11]: Jsdoc

File: src/Component/HttpFoundation/src/Websocket/Websocket.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+const Frame = Jymfony.Component.HttpFoundation.Websocket.Frame;
+const Message = Jymfony.Component.HttpFoundation.Websocket.Message;
+
+/**
+ * Base class for websocket handlers.
+ *
+ * @memberOf Jymfony.Component.HttpFoundation.Websocket
+ */
+class Websocket {
+    __construct() {
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/Websocket.js:10]: jsdoc

File: src/Component/HttpFoundation/src/Websocket/WebsocketResponse.js [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+     * @inheritdoc
+     */
+    prepare(request) {
+        super.prepare(request);
+
+        this.headers.set('Upgrade', 'websocket');
+        this.headers.set('Connection', 'upgrade');
+
+        const sha1 = crypto.createHash('sha1');
+        sha1.update(request.headers.get('Sec-WebSocket-Key') + '258EAFA5-E914-47DA-95CA-C5AB0DC85B11');
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Websocket/WebsocketResponse.js]: Can maybe this guid set into a constant of `WebsocketResponse`? The constant can be named with something like `RFC6455_GUID`

File: src/Component/HttpServer/src/EventListener/WebsocketListener.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            return;
+        }
+
+        if (! request.headers.has('Sec-WebSocket-Version')) {
+            throw new HttpException(Response.HTTP_BAD_REQUEST, 'Client must provide a value for Sec-WebSocket-Version.');
+        }
+
+        const websocketVersion = ~~request.headers.get('Sec-WebSocket-Version');
+        switch (websocketVersion) {
+            case 8:
```
Comment (massimilianobraglia) [inline src/Component/HttpServer/src/EventListener/WebsocketListener.js:34]: Using constants here can be useful and more readable

**Review by massimilianobraglia — CHANGES_REQUESTED** [review]

File: src/Component/Debug/src/Exception/UnhandledRejectionException.js (hunk 3 above)
Comment (alekitto) [reply src/Component/Debug/src/Exception/UnhandledRejectionException.js]: I've entirely changed the construction of this exception message. Now should be more developer-friendly, showing the user the previous error if any, or signaling that a non-Error object has been used to reject the promise

File: src/Component/HttpFoundation/src/Websocket/FrameBuffer.js (hunk 4 above)
Comment (alekitto) [reply src/Component/HttpFoundation/src/Websocket/FrameBuffer.js]: Exception added

**Comment by alekitto** [comment]
All done.

**Pull request merged**